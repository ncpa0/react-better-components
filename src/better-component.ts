import type { ReactNode } from "react";
import React from "react";
import { Effect } from "./modules/effect";
import { LifecycleEvents } from "./modules/lifecycle";
import type { PublicOf, Reducer } from "./modules/reducer";
import { createReducerState } from "./modules/reducer";
import { State } from "./modules/state";

export abstract class BetterComponent<
  P extends object = {},
> extends React.Component<P> {
  public state: Record<string, any> = {};

  private _nextStateId = 1;
  private _lifecycle = new LifecycleEvents();

  constructor(props: Readonly<P> | P) {
    super(props);

    const injectedContexts = Object.entries(props).filter(([k]) =>
      k.startsWith("__ctx_"),
    );

    for (let i = 0; i < injectedContexts.length; i++) {
      const [key, value] = injectedContexts[i]!;
      const fieldName = key.slice("__ctx_".length);

      Object.defineProperty(this, fieldName, {
        configurable: true,
        enumerable: true,
        value: value,
      });

      let lastCtxValue = value;
      this._lifecycle.onUpdate(() => {
        // @ts-expect-error
        const currentCtxValue = this.props[key];
        if (currentCtxValue !== lastCtxValue) {
          Object.defineProperty(this, fieldName, {
            configurable: true,
            enumerable: true,
            value: currentCtxValue,
          });
          this.forceUpdate();
        }
        lastCtxValue = currentCtxValue;
      });
    }
  }

  public componentDidMount(): void {
    this._lifecycle.mount();
  }

  public componentDidUpdate(): void {
    this._lifecycle.update();
  }

  public componentWillUnmount(): void {
    this._lifecycle.unmount();
  }

  protected $state<T>(initialValue: T | (() => T)): State<T>;
  protected $state<T = undefined>(): State<T | undefined>;
  protected $state<T>(initialValue?: T) {
    return new State(
      this._lifecycle,
      String(this._nextStateId++),
      () => this.state,
      (s) => this.setState(s),
      initialValue,
    );
  }

  protected $reducer<A extends object, T = undefined>(
    actionsClass: new () => A,
  ): Reducer<T, A>;
  protected $reducer<A extends object, T>(
    initialValue: T,
    actionsClass: new () => A,
  ): Reducer<T, A>;
  protected $reducer(...args: any[]) {
    const initialValue = args.length === 1 ? undefined : args[0];
    const actionsClass = args.length === 1 ? args[0] : args[1];

    return createReducerState(
      this._lifecycle,
      String(this._nextStateId++),
      () => this.state,
      (s) => this.setState(s),
      actionsClass,
      initialValue,
    );
  }

  protected $effect(
    callback: () => void | (() => void),
    deps?: PublicOf<State<any>>[],
  ) {
    return new Effect(this._lifecycle, callback, deps?.slice());
  }

  abstract render(): ReactNode;
}
