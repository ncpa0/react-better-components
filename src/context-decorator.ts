/* eslint-disable @typescript-eslint/no-unused-vars */
import React from "react";

export const MapContext =
  (context: Record<string, React.Context<any>>) =>
  (Component: any, _: any) => {
    const ctxEntries = Object.entries(context);

    const getJsxElem = ctxEntries.reduce(
      (
        elem: (additionalProps: Record<string, any>) => JSX.Element,
        [name, ctx],
      ) => {
        return (additionalProps: Record<string, any>) =>
          React.createElement(
            ctx.Consumer,
            undefined,
            // @ts-expect-error
            (ctxValue: any) => {
              additionalProps["__ctx_" + name] = ctxValue;
              return elem(additionalProps);
            },
          );
      },
      (additionalProps: Record<string, any>) =>
        React.createElement(Component, additionalProps),
    );

    return class ContextInjector extends React.Component {
      render() {
        return getJsxElem({ ...this.props });
      }
    } as any;
  };
