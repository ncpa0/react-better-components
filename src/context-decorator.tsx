/* eslint-disable @typescript-eslint/no-unused-vars */
import React from "react";

export const MapContext =
  (context: Record<string, React.Context<any>>) =>
  (Component: any, _?: any) => {
    const ctxEntries = Object.entries(context);

    const getJsxElem = ctxEntries.reduce(
      (
        elem: (additionalProps: Record<string, any>) => JSX.Element,
        [name, ctx],
      ) => {
        return (additionalProps: Record<string, any>) => (
          <ctx.Consumer>
            {(ctxValue: any) => {
              additionalProps["__ctx_" + name] = ctxValue;
              return elem(additionalProps);
            }}
          </ctx.Consumer>
        );
      },
      (additionalProps: Record<string, any>) => (
        <Component {...additionalProps} />
      ),
    );

    return class ContextInjector extends React.Component {
      render() {
        return getJsxElem({ ...this.props });
      }
    } as any;
  };
