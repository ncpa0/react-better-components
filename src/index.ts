export { BetterComponent } from "./better-component";
export { MapContext } from "./context-decorator";
export type { Effect } from "./modules/effect";
export type { Reducer, ReducerActionsOf } from "./modules/reducer";
export type { State } from "./modules/state";
