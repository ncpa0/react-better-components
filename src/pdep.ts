import type { Dependency } from "./modules/effect";

/**
 * Creates a dependency from a component property that can be used in
 * an effect.
 *
 * @example
 *   this.$effect(() => {}, [pdep(() => this.props.label)]);
 */
export const pdep = <T>(getProp: () => T): Dependency<T> => {
  return {
    get: getProp,
  };
};
