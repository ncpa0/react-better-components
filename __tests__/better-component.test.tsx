import {
  act,
  fireEvent,
  render,
  waitFor,
} from "@testing-library/react";
import React from "react";
import { BetterComponent } from "../src/index";

type ButtonProps = {
  onClick?: () => void;
  onLabelChange?: (label: string) => void;
};

class Button extends BetterComponent<ButtonProps> {
  private label = this.$state("Click Me");
  private className = this.$state("btn btn-primary");

  constructor(props: ButtonProps) {
    super(props);

    this.$effect(() => {
      this.props.onLabelChange?.(this.label.get());
    }, [this.label]);
  }

  handleClick = () => {
    this.label.set("Clicked");
    this.props.onClick?.();
  };

  render() {
    return (
      <button
        data-testid="comp-btn"
        className={this.className.get()}
        onClick={this.handleClick}
      >
        {this.label}
      </button>
    );
  }
}

describe("BetterComponent", () => {
  it("should render the component", async () => {
    const app = render(
      <div>
        <Button />
      </div>,
    );

    expect(app.asFragment()).toMatchSnapshot();
  });

  it("should trigger the effects", async () => {
    const onBtnLabelChange = jest.fn();

    const app = render(
      <div>
        <Button onClick={() => {}} onLabelChange={onBtnLabelChange} />
      </div>,
    );

    const button = await app.findByTestId("comp-btn");

    expect(button).toBeDefined();
    expect(button.textContent).toBe("Click Me");
    expect(button.className).toBe("btn btn-primary");

    fireEvent.click(button);

    await waitFor(async () => {
      const button = await app.findByTestId("comp-btn");
      expect(button.textContent).toBe("Clicked");
      expect(onBtnLabelChange).toHaveBeenCalledWith("Clicked");
    });
  });

  it("should correctly handle reducer states", async () => {
    class NumberReducer {
      increment(state: number) {
        return state + 1;
      }

      decrement(state: number) {
        return state - 1;
      }

      multiply(state: number, payload: number) {
        return state * payload;
      }
    }

    const onCounterChange = jest.fn();

    class Counter extends BetterComponent {
      c = this.$reducer(0, NumberReducer);

      constructor(props: {}) {
        super(props);

        this.$effect(() => {
          onCounterChange(this.c.get());
        }, [this.c]);
      }

      render() {
        return (
          <div>
            <button
              data-testid="inc-btn"
              onClick={() => {
                this.c.increment();
              }}
            >
              Increment
            </button>
            <button
              data-testid="dec-btn"
              onClick={() => {
                this.c.decrement();
              }}
            >
              Decrement
            </button>
            <button
              data-testid="mul-btn"
              onClick={() => {
                this.c.multiply(2);
              }}
            >
              Multiply
            </button>
            <div data-testid="counter-out">{this.c}</div>
          </div>
        );
      }
    }

    const App = () => {
      return (
        <>
          <Counter />
        </>
      );
    };

    const rendered = render(<App />);

    const getCurrentCounter = () => {
      return rendered.getByTestId("counter-out").textContent;
    };

    const clickIncrement = () => {
      fireEvent.click(rendered.getByTestId("inc-btn"));
    };

    const clickDecrement = () => {
      fireEvent.click(rendered.getByTestId("dec-btn"));
    };

    const clickMultiply = () => {
      fireEvent.click(rendered.getByTestId("mul-btn"));
    };

    expect(getCurrentCounter()).toBe("0");

    clickIncrement();

    await waitFor(() => {
      expect(getCurrentCounter()).toBe("1");
    });

    clickIncrement();

    await waitFor(() => {
      expect(getCurrentCounter()).toBe("2");
    });

    act(() => {
      clickIncrement();
      clickIncrement();
      clickIncrement();
    });

    await waitFor(() => {
      expect(getCurrentCounter()).toBe("5");
    });

    clickDecrement();

    await waitFor(() => {
      expect(getCurrentCounter()).toBe("4");
    });

    clickMultiply();

    await waitFor(() => {
      expect(getCurrentCounter()).toBe("8");
    });

    expect(onCounterChange).toHaveBeenCalledTimes(6);
    expect(onCounterChange.mock.calls).toEqual([
      [0],
      [1],
      [2],
      [5],
      [4],
      [8],
    ]);
  });
});
