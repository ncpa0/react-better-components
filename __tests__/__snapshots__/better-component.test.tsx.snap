// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetterComponent should render the component 1`] = `
<DocumentFragment>
  <div>
    <button
      class="btn btn-primary"
      data-testid="comp-btn"
    >
      Click Me
    </button>
  </div>
</DocumentFragment>
`;
